addon/services/file-manager.js

```diff
@@ -0,0 +1,114 @@
+            kind: 'file'
+        }, contents);
+    },
+
+    checkout(file, user) {
+        // TODO? Having checkout here makes more sense to me than making it
+        // the only writable attribute on the file model.
+    },
+
+    // Folder actions
+    addSubfolder(folder, name) {
+        var url = folder.get('links').new_folder;
+
+        // HACK: This is the only link that already has a query string
+        if (url.search(/\?kind=folder$/) > -1) {
+            url = `${url}&name=${name}`;
+            return this._waterbutlerRequest('PUT', url);
+        } else {
+            return this._waterbutlerRequest('PUT', url, {
+                name,
+                kind: 'folder'
+            });
+        }
+    },
+
+    uploadFile(folder, name, contents) {
+        var url = folder.get('links').upload;
+        return this._waterbutlerRequest('PUT', url, {
+            name,
+            kind: 'file'
+        }, contents);
+    },
+
+    // File and folder actions
+    rename(file, newName) {
+        var url = file.get('links').move;
+        return this._waterbutlerRequest('POST', url, data = {
+            action: 'rename',
+            rename: newName
+        });
+    },
+
+    move(file, targetFolder, newName=null, replace=true, node=null,
+            provider=null, action='move') {
+        var url = file.get('links').move;
+        var data = {
+            action,
+            path: targetFolder.get('path'),
+            conflict: replace ? 'replace' : 'keep'
+        };
+        if (newName) {
+            data.rename = newName;
+        }
+        if (node) {
+            if (typeof node === 'string') {
+                data.resource = node;
+            } else {
+                data.resource = node.get('id');
+            }
+        }
+        if (provider) {
+            data.provider = provider;
+        }
+        return this._waterbutlerRequest('POST', url, data=data);
+    },
+
+    copy(file, targetFolder, newName=null,
+            replace=true, node=null, provider=null) {
+        return this.move(file, targetFolder, newName, replace,
+                node, provider, 'copy');
+    },
+
+    deleteFile(file) {
+        var url = folder.get('links').delete;
+        return this._waterbutlerRequest('DELETE', url);
+    },
+
+    _waterbutlerRequest(method, url, queryParams=null, data=null) {
+        if (queryParams) {
+            let queryString = Ember.$.param(queryParams);
+            url = `${url}?${queryString}`;
+        }
+        var sessionData = this.get('session').get('data');
+        var accessToken = sessionData.authenticated.attributes.accessToken;
+        return new Ember.RSVP.Promise(function(resolve, reject) {
+            Ember.$.ajax(url, {
+                method,
+                headers: {
+                    Authorization: `Bearer ${accessToken}`
+                },
+                error(_, __, errorMsg) {
+                    reject(new Error(errorMsg));
```

Comment:
Might be good to return a DS.Errors instance?
http://emberjs.com/api/data/classes/DS.Errors.html

thoughts?